psl-core/src/main/java/org/linqs/psl/util/VizDataCollection.java

```diff
@@ -0,0 +1,233 @@
+            Iterable<GroundRule> groundedRuleList = groundRuleStore.getGroundRules(rule);
+            double totalSat = 0.0;
+            double totalDis = 0.0;
+            int groundRuleCount = 0;
+            JSONObject moduleElement = new JSONObject();
+
+            for (GroundRule groundRule : groundedRuleList) {
+                if (groundRule instanceof WeightedGroundRule) {
+                    WeightedGroundRule weightedGroundRule = (WeightedGroundRule)groundRule;
+                    totalSat += 1.0 - weightedGroundRule.getIncompatibility();
+                    totalDis += weightedGroundRule.getIncompatibility();
+                }
+                groundRuleCount++;
+            }
+            moduleElement.put("Rule", rule.getName());
+            moduleElement.put("Total Satisfaction", totalSat);
+            moduleElement.put("Satisfaction Percentage", totalSat / groundRuleCount);
+            moduleElement.put("Total Dissatisfaction", totalDis);
+            moduleElement.put("Dissatisfaction Percentage", totalDis / groundRuleCount);
+            vizData.totRuleSatArray.put(moduleElement);
+        }
+    }
+
+    public static void violatedGroundRules(List<Rule> rules, GroundRuleStore groundRuleStore) {
+        for (Rule rule : rules) {
+            JSONObject moduleElement = new JSONObject();
+            double violation = 0.0;
+            boolean weightFlag = false;
+            Iterable<GroundRule> groundedRuleList = groundRuleStore.getGroundRules(rule);
+            for (GroundRule groundRule : groundedRuleList) {
+                if (vizData.violatedGroundRulesList.contains(groundRule)) {
+                    //There can't be weighted violated rules so we can make an assumption here
+                    UnweightedGroundRule unweightedGroundRule = (UnweightedGroundRule)groundRule;
+                    violation = unweightedGroundRule.getInfeasibility();
+                    moduleElement.put("Violated Rule", groundRule.baseToString());
+                    moduleElement.put("Parent Rule", rule.getName());
+                    // moduleElement.put("Weighted", weightFlag);
+                    moduleElement.put("Violation", violation);
+                    vizData.violatedGroundRulesArray.put(moduleElement);
+                }
+            }
+        }
+    }
+
+    public static void ruleMapInsertElement(AbstractLogicalRule parentRule, GroundRule groundRule,
+                            Map<Variable, Integer> variableMap,  Constant[] constantsList) {
+      //Adds a groundAtom element to RuleMap
+      //Why are some groundRules null? Perhaps a thread thing??
+      ArrayList<Integer> atomHashList = new ArrayList<>();
+      if (groundRule != null) {
+        HashSet<Atom> atomSet = new HashSet<>(groundRule.getAtoms());
+        int atomCount = 0;
+        HashMap<String,String> atomMap = new HashMap<>();
+        for (Atom a : atomSet) {
+          atomHashList.add(System.identityHashCode(a));
```

Comment:
Why do you immediately convert the object id into a String?
Why not leave as an Integer?

Reply:
The line this comment is on has no string conversion. Are you talking about this line: https://github.com/AaronRodden/psl/blob/psl_viz/psl-core/src/main/java/org/linqs/psl/util/VizDataCollection.java#L117

I converted it because JSON's require strings as keys so I thought I would just do the conversion there.

Reply:
The concern here is not that you convert it to a String in general, but that it is done here.
Conversions like this (where you have to make it because the data format requires it) should generally be done at the edges (right when it is necessary).